### PR TITLE
Add Horovod, AMP and DALI to SSD training script

### PR DESCRIPTION
**Pull request by Kh4L**
Add three options to accelerate training of SSD with COCO:
- `--dali` Use [DALI](https://docs.nvidia.com/deeplearning/sdk/dali-developer-guide/docs/index.html) for faster data loading and data preprocessing in training with COCO dataset. DALI >= 0.11 required.
- `--amp` Use [Automatic Mixed Precision training](https://mxnet.incubator.apache.org/versions/master/tutorials/amp/amp_tutorial.html), automatically casting FP16 where safe
- `--horovod` Use [Horovod](https://github.com/horovod/horovod) for distributed training, with a network agnostic wrapper for the optimizer, allowing efficient allreduce using OpemMPI and NCCL

All combined, these changes yield a ~4.5x speed improvement on the DGX-1V (8xTesla V100 GPUs) it was tested on:

```
python train_ssd.py --dataset coco -j 4 --gpus 0,1,2,3,4,5,6,7 --network resnet50_v1 --data-shape 512 --log-interval 5
...
INFO:root:[Epoch 0][Batch 2004], Speed: 34.658 samples/sec, CrossEntropy=5.614, SmoothL1=3.033
INFO:root:[Epoch 0][Batch 2009], Speed: 40.668 samples/sec, CrossEntropy=5.612, SmoothL1=3.032
INFO:root:[Epoch 0][Batch 2014], Speed: 124.791 samples/sec, CrossEntropy=5.610, SmoothL1=3.031
INFO:root:[Epoch 0][Batch 2019], Speed: 126.232 samples/sec, CrossEntropy=5.607, SmoothL1=3.031
INFO:root:[Epoch 0][Batch 2024], Speed: 105.187 samples/sec, CrossEntropy=5.606, SmoothL1=3.030
INFO:root:[Epoch 0][Batch 2029], Speed: 28.480 samples/sec, CrossEntropy=5.604, SmoothL1=3.029
```

```
horovodrun -np 8 -H localhost:8 python ./gluon-cv/scripts/detection/ssd/train_ssd.py --dataset coco -j 4 --network resnet50_v1 --data-shape 512 --log-interval 5 --horovod --dali --amp
...
INFO:root:[Epoch 0][Batch 2004], Speed: 316.286 samples/sec, CrossEntropy=5.601, SmoothL1=2.271
INFO:root:[Epoch 0][Batch 2009], Speed: 332.049 samples/sec, CrossEntropy=5.598, SmoothL1=2.270
INFO:root:[Epoch 0][Batch 2014], Speed: 347.680 samples/sec, CrossEntropy=5.595, SmoothL1=2.269
INFO:root:[Epoch 0][Batch 2019], Speed: 315.211 samples/sec, CrossEntropy=5.594, SmoothL1=2.268
INFO:root:[Epoch 0][Batch 2024], Speed: 332.870 samples/sec, CrossEntropy=5.590, SmoothL1=2.267
INFO:root:[Epoch 0][Batch 2029], Speed: 332.070 samples/sec, CrossEntropy=5.587, SmoothL1=2.266
```

Even though the training localisation loss (SmoothL1) differs, the validation mAP is similar:

convergence_curve_ssd_rn50_mxnet

Signed-off-by: Serge Panev <spanev@nvidia.com>

### CODE REVIEW

**Comment by mli** [comment]
Job PR-807-2 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-807/2/index.html 
Code coverage of this PR:  vs. Master:

File: gluoncv/data/transforms/presets/ssd.py [hunk 1]
```diff
@@ -211,3 +221,128 @@ def __call__(self, src, label):
         img = mx.nd.image.to_tensor(img)
         img = mx.nd.image.normalize(img, mean=self._mean, std=self._std)
         return img, bbox.astype(img.dtype)
+
+class SSDCocoDALIPipeline(Pipeline):
+    """DALI Pipeline with COCO Reader and SSD training transorm.
```
Comment (Kh4L) [inline gluoncv/data/transforms/presets/ssd.py]:
```suggestion
    """DALI Pipeline with COCO Reader and SSD training transform.
```
Waiting for further review before committing the change

File: gluoncv/data/transforms/presets/ssd.py [hunk 2]
```diff
@@ -211,3 +221,128 @@ def __call__(self, src, label):
...
+
+    Parameters
+    ----------
+    num_shards: int
+         DALI pipeline arg - Number of pipelines used indicates to the reader
+         how to split/shard the dataset.
+    device_id: int
+         DALI pipeline arg - Device id.
+    shard_id: int
+         DALI pipeline arg - Shard id of the pipeline Must be in [0, num_shards).
```
Comment (Kh4L) [inline gluoncv/data/transforms/presets/ssd.py]:
```suggestion
         DALI pipeline arg - Shard id of the pipeline must be in [0, num_shards).
```
Same here

File: gluoncv/data/transforms/presets/ssd.py [hunk 3]
```diff
@@ -211,3 +221,128 @@ def __call__(self, src, label):
...
         img = mx.nd.image.normalize(img, mean=self._mean, std=self._std)
         return img, bbox.astype(img.dtype)
+
+class SSDCocoDALIPipeline(Pipeline):
+    """DALI Pipeline with COCO Reader and SSD training transorm.
+
+    Parameters
+    ----------
+    num_shards: int
+         DALI pipeline arg - Number of pipelines used indicates to the reader
```
Comment (Kh4L) [inline gluoncv/data/transforms/presets/ssd.py]:
```suggestion
         DALI pipeline arg - Number of pipelines used indicating to the reader
```

**Comment by ptrendx** [comment]
@zhreshold

File: scripts/detection/ssd/train_ssd.py [hunk 4]
```diff
@@ -105,6 +129,72 @@ def get_dataloader(net, train_dataset, val_dataset, data_shape, batch_size, num_
...
+        anchors = anchors.as_in_context(mx.cpu())
+
+        # prepare anchors into ltrb (normalized DALI anchors format list)
+        anchors_ltrb = to_normalized_ltrb_list(anchors, data_shape)
+
+        # training
+        expanded_file_root = os.path.expanduser(file_root)
+        coco_root = expanded_file_root + '/coco/train2017'
+        coco_annotations = expanded_file_root + '/coco/annotations/instances_train2017.json'
+        coco_size = 118287
```
Comment (zhreshold) [inline scripts/detection/ssd/train_ssd.py]: how can I get coco_size in a non-hardcoded way?

File: gluoncv/data/transforms/presets/ssd.py [hunk 5]
```diff
@@ -211,3 +221,128 @@ def __call__(self, src, label):
         img = mx.nd.image.to_tensor(img)
         img = mx.nd.image.normalize(img, mean=self._mean, std=self._std)
         return img, bbox.astype(img.dtype)
+
+class SSDCocoDALIPipeline(Pipeline):
```
Comment (zhreshold) [inline gluoncv/data/transforms/presets/ssd.py]: can we decouple dataset(COCO/VOC) with algorithm(SSD DALI graph)?

File: scripts/detection/ssd/train_ssd.py [hunk 6]
```diff
@@ -105,6 +129,72 @@ def get_dataloader(net, train_dataset, val_dataset, data_shape, batch_size, num_
         batch_size, False, batchify_fn=val_batchify_fn, last_batch='keep', num_workers=num_workers)
     return train_loader, val_loader
 
+def to_normalized_ltrb_list(anchors, size):
```
Comment (zhreshold) [inline scripts/detection/ssd/train_ssd.py]: we could move this util function into SSD DALI pipeline

File: scripts/detection/ssd/train_ssd.py [hunk 7]
```diff
@@ -105,6 +129,72 @@ def get_dataloader(net, train_dataset, val_dataset, data_shape, batch_size, num_
...
+    anchors_np = anchors.squeeze().asnumpy()
+    anchors_np_ltrb = anchors_np.copy()
+    anchors_np_ltrb[:, 0] = anchors_np[:, 0] - 0.5 * anchors_np[:, 2]
+    anchors_np_ltrb[:, 1] = anchors_np[:, 1] - 0.5 * anchors_np[:, 3]
+    anchors_np_ltrb[:, 2] = anchors_np[:, 0] + 0.5 * anchors_np[:, 2]
+    anchors_np_ltrb[:, 3] = anchors_np[:, 1] + 0.5 * anchors_np[:, 3]
+    anchors_np_ltrb /= size
+    return anchors_np_ltrb.flatten().tolist()
+
+def get_dali_dataloader(net, dataset_name, data_shape, global_batch_size, num_workers, devices, file_root, ctx, horovod):
```
Comment (zhreshold) [inline scripts/detection/ssd/train_ssd.py]: here in the script, there are too many hardcoded paths, similar to gluoncv's original dataset, I think DALI dataset reader should also follow the same tradition, e.g., in https://github.com/dmlc/gluon-cv/tree/master/gluoncv/data/mscoco/detection.py , we can have a `DALICOCODetection` class which handles the dataset location and `coco_size`. It will be shared by all detection algorithms.

Comment (Kh4L) [reply scripts/detection/ssd/train_ssd.py]: @zhreshold this and previews comments addressed in https://github.com/dmlc/gluon-cv/pull/807/commits/7be871ea66de988055cf1359d1c831a0bd532f7c

**Comment by mli** [comment]
Job PR-807-4 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-807/4/index.html 
Code coverage of this PR:  vs. Master:

File: gluoncv/data/mscoco/detection.py [hunk 8]
```diff
@@ -195,3 +203,45 @@ def _check_load_bbox(self, coco, entry):
...
+        """Returns three DALI graph nodes: inputs, bboxes, labels.
+        To be called in `define_graph`.
+        """
+        inputs, bboxes, labels = self.input(name="Reader")
+        images = self.decode(inputs)
+        return (images, bboxes, labels)
+    def size(self):
+        """Returns size of COCO dataset
+        """
+        coco_size = 118287
```
Comment (zhreshold) [inline gluoncv/data/mscoco/detection.py]: anyway to parse the annotation_file? we have users who use their custom dataset in coco compatible format

Comment (zhreshold) [reply gluoncv/data/mscoco/detection.py]: is COCOReader able to return the number of images?

File: gluoncv/data/transforms/presets/ssd.py [hunk 9]
```diff
@@ -6,7 +6,17 @@
 from .. import image as timage
 from .. import experimental
 
-__all__ = ['transform_test', 'load_test', 'SSDDefaultTrainTransform', 'SSDDefaultValTransform']
+try:
```
Comment (zhreshold) [inline gluoncv/data/transforms/presets/ssd.py]: can we move duplicate dali import utils into https://github.com/dmlc/gluon-cv/blob/49056d4892382fca067216d6271d1633d869cb8d/gluoncv/utils/filesystem.py?

File: scripts/detection/ssd/train_ssd.py [hunk 10]
```diff
@@ -105,6 +129,72 @@ def get_dataloader(net, train_dataset, val_dataset, data_shape, batch_size, num_
         batch_size, False, batchify_fn=val_batchify_fn, last_batch='keep', num_workers=num_workers)
     return train_loader, val_loader
 
+def get_dali_dataset(dataset_name, devices, args):
+    if dataset_name.lower() == "coco":
+        # training
+        expanded_file_root = os.path.expanduser(args.dataset_root)
+        coco_root = expanded_file_root + '/coco/train2017'
```
Comment (zhreshold) [inline scripts/detection/ssd/train_ssd.py]: `os.path.join`

**Comment by hetong007** [comment]
Not everyone knows how to configure for Horovod or AMP. It is the best to have this script with a tutorial as an introduction.

File: gluoncv/data/mscoco/detection.py (hunk 8 above)
Comment (Kh4L) [reply gluoncv/data/mscoco/detection.py]: I see, yes COCOReader is able to return it, but we have to build the whole DALI pipeline then destroy i
It's not really clean.. but would be actually faster than the default Python parser that takes 15s to parse vs 0.5s for DALI.

Comment (Kh4L) [reply gluoncv/data/mscoco/detection.py]: One feature on DALI roadmap is to have standalone ops that can be directly run in Python, in the meantime I'll add this "1-op micro Pipeline"

**Comment by Kh4L** [comment]
@zhreshold @hetong007 sure, I will add some info in the README. I think pointing on the official tutorials would be the best here

Another thing would be to have a tutorial in the form of https://gluon-cv.mxnet.io/build/examples_detection/train_ssd_voc.html 
Something like "Accelerate the training of SSD with AMP, Horovod and DALI", as a 11th section of the Object Detection tutorial.
This might be part of another PR?

**Comment by hetong007** [comment]
Another PR explaining it is great. Besides, do you plan to merge these options to other training scripts as well?

**Comment by Kh4L** [comment]
Yes, I am working on adding them to the Faster-RCNN and YOLO training scripts

File: gluoncv/data/mscoco/detection.py (hunk 8 above)
Comment (Kh4L) [reply gluoncv/data/mscoco/detection.py]: This and other review comments addressed in https://github.com/dmlc/gluon-cv/pull/807/commits/3b6258c4d2fa06cb304d14acf395abfe1620b46e

**Comment by mli** [comment]
Job PR-807-6 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-807/6/index.html 
Code coverage of this PR:  vs. Master:

**Comment by zhreshold** [comment]
looks good now, I will try training with dali/amp/hovorod on my side and will merge this PR once done.

**Comment by mli** [comment]
Job PR-807-7 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-807/7/index.html 
Code coverage of this PR:  vs. Master:

**Comment by mli** [comment]
Job PR-807-8 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-807/8/index.html 
Code coverage of this PR:  vs. Master:

**Comment by zhreshold** [comment]
lgtm, @Kh4L let me know if you think it's ready for merge

**Comment by mli** [comment]
Job PR-807-9 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-807/9/index.html 
Code coverage of this PR:  vs. Master:

**Comment by Kh4L** [comment]
@zhreshold DALI 0.11 is live, ready to merge

**Comment by zhreshold** [comment]
@Kh4L Thanks, this is merged! Could you add a tutorial regarding all the new techniques, especially the setups. Myself found there are quite a few pitfalls to avoid, especially the hovorod and NCCL installation.

File: scripts/detection/ssd/train_ssd.py [hunk 11]
```diff
@@ -253,11 +388,25 @@ def train(net, train_data, val_data, eval_metric, ctx, args):
             warnings.simplefilter("always")
             net.initialize()
             async_net.initialize()
+            # needed for net to be first gpu when using AMP
+            net.collect_params().reset_ctx(ctx[0])
```
Comment (jpizarrom) [inline scripts/detection/ssd/train_ssd.py]: I got higher mAP when I am not using amp and removing this line,
I got lower mAP when I am not using amp and keep this line,
cc @zhreshold @Kh4L

Comment (Kh4L) [reply scripts/detection/ssd/train_ssd.py]: How big is the mAP difference?

Comment (jpizarrom) [reply scripts/detection/ssd/train_ssd.py]: 8% mAP in the first epoch with my own dataset. I think it is related to the GPU initialization with the empty data used to obtain the anchors.

I was able to obtain comparable mAP on the first epoch, I had reloaded the entire network after get_dataloader, calling get_model and initialize again.

Comment (Kh4L) [reply scripts/detection/ssd/train_ssd.py]: When you leave the line the anchors are first generated by the model on GPU 0 then copied to CPU, but when comment out this line, you are using anchors generated directly the model on the CPU. Both should produce an identical result.

> I was able to obtain comparable mAP on the first epoch, I had reloaded the entire network after get_dataloader, calling get_model and initialize again.

You managed to get a higher mAP with `net.collect_params().reset_ctx(ctx[0])`, but you had to add lines to reload the whole model after `get_dataloader`. Did I understand correctly?

Comment (jpizarrom) [reply scripts/detection/ssd/train_ssd.py]: Yes,

I managed to get a higher mAP with `net.collect_params().reset_ctx(ctx[0])`, but I had to add lines to reload the whole model after `get_dataloader`.

It is strange, I was expecting to obtain the same results.

Comment (zhreshold) [reply scripts/detection/ssd/train_ssd.py]: @PariksheetPinjari909  I think it's weird according to your description. Will you be able to compare the anchors generated w/without removing this line?

Comment (jpizarrom) [reply scripts/detection/ssd/train_ssd.py]: The anchors are the same, maybe It could be related to the weight initialization.

It looks like the issue occurs when a resnet is used as backbone such as ssd_512_resnet50_v1_custom.

I will try to prepare a code sample.

**Comment by austinmw** [comment]
Would love a tutorial for running Horovod+DALI+AMP. Also any progress on adding it to other train scripts?
